test_analyzer_streamline_single.py

```diff
@@ -0,0 +1,226 @@
+hist4 = r.TH2F("2e_tag_aper_prob_100R_v2","Title;Y (barID);F(Y)",50,0,50,50,0,5)
+hist5 = r.TH2F("2e_up_aper_prob_100R_v2","Title;Y (barID);F(Y)",50,0,50,50,0,5)
+hist6 = r.TH2F("2e_down_aper_prob_100R_v2","Title;Y (barID);F(Y)",50,0,50,50,0,5)
+hists_aper = [hist4, hist5, hist6]
+if (hist_dim == '2D'):
+	hist7 = r.TH2F("2e_tag_" + variable + "_" + variable2 + "_100R_v2","Title;" + x_ax + ";" + y_ax,xbins,xlow,xhigh,ybins,ylow,yhigh)
+	hist8 = r.TH2F("2e_up_" + variable + "_" + variable2 + "_100R_v2","Title;" + x_ax + ";" + y_ax,xbins,xlow,xhigh,ybins,ylow,yhigh)
+	hist9 = r.TH2F("2e_down_" + variable + "_" + variable2 + "_100R_v2","Title;" + x_ax + ";" + y_ax,xbins,xlow,xhigh,ybins,ylow,yhigh)
+	hists = [hist7, hist8, hist9]
+
+#record relative hit efficiency and secondary parameters
+heven_r = [0,0,0]
+hodd_r = [0,0,0]
+heven_f = [0,0,0]
+hodd_f = [0,0,0]
+entries_r = [0,0,0] 	#scale factor is configured in simulation generation script when user determines number of e-/event
+entries_f=[0,0,0]
+
+#cumulative calculations of EEP per module
+#hit_dict_tag = {}
+#hit_dict_up = {}
+#hit_dict_down ={}
+#hit_dicts = [hit_dict_tag, hit_dict_up,hit_dict_down]
+#for i in range(len(hit_dicts)):
+#	for j in range(50):
+#		hit_dicts[i][j] = 0
+
+
+def aper_ratio(x, next_x, prev_x):
+        return (float(next_x + prev_x)/x)
+
+
+## loop over events
+for i in range(cont.tree.numentries):
+    for j in range(len(modules)) :
+    	## get list of pe for tagger array for event i
+    	params = cont.get_data(modules[j],variable,i)
+	if (hist_dim == "2D"):
+		params2 = cont.get_data(modules[j],variable2,i)
+    	barNums=cont.get_data(modules[j],'barID',i)
+    	noiseDiscr = cont.get_data(modules[j],'isNoise',i)
+	rnoiseDiscr = cont.get_data(modules[j],'beamEfrac',i)
+	hit_dict = {}
+	for l in range(50):
+		hit_dict[l] = 0
+    	for k in range(len(params)) :  
+		if ((hist_dim == '1D') and ((pNoise.lower() == 'n') or ((pNoise.lower() == 'y') and (noiseDiscr[k] == False))) and ((sNoise.lower() == 'n') or ((sNoise.lower() == 'y') and ((rnoiseDiscr[k] < 0.2) or (rnoiseDiscr[k] > 0.9)) ))):
+        		hists[j].Fill(params[k])
+		if ((hist_dim == '2D') and ((pNoise.lower() == 'n') or ((pNoise.lower() == 'y') and (noiseDiscr[k] == False))) and ((sNoise.lower() == 'n') or ((sNoise.lower() == 'y') and ((rnoiseDiscr[k] < 0.2) or (rnoiseDiscr[k] > 0.9)) ))):
+			hists[j].Fill(params[k],params2[k])
+		hit_dict[barNums[k]]+=1
+		if (rnoiseDiscr[k] > 0.9):
+			entries_r[j]+=1
+			if (barNums[k] % 2 == 0):
+                		heven_r[j]+=1
+        		else:
+                		hodd_r[j]+=1
+		if (rnoiseDiscr[k] <= 0.9):
+			entries_f[j]+=1	
+        		if (barNums[k] % 2 == 0):
+				heven_f[j]+=1
+			else:
+				hodd_f[j]+=1
+	if(aper_tog.lower() == 'y'):
+		bars = hit_dict.keys()
+		hits = hit_dict.values()
+		for n in range(1,len(bars)-1):
+			if ((n % 2 != 0) and (hits[n] != 0)):
+				hists_aper[j].Fill(bars[n],aper_ratio(hits[n],hits[n+1],hits[n-1]))
```

Comment:
here i don't see a check on the actual barID, just that you're passing the neighboring elements. looking earlier in the loop, i see that for each k = 1, ..., N_elements_in_param, hit_dict[barNum[k]] gets incremented. 
so, if we have 5 hits at barIDs 2, 3, 6, 14, 28, then do we have an array hits which is 50 elements, most of them 0, except for at those five indices? in that case it's fine. if hits indeed only has 5 elements, then this logic doesn't work.

Reply:
Yeah it's exactly that, an array of 50 elements. That's why I have an "initialization" loop earlier (the one that uses the "l' iterator). It fills the dictionary with 0 hit values corresponding each barID 0-49, so the hits array will be a 50 element array, predominantly sparse.

Reply:
That's also why I included the hits[n] != 0 condition in the if statement, because some hits values will be and we want to avoid dividing by zero when calculating the EEP